deployer/src/deployment/queue.rs

```diff
@@ -79,3 +122,169 @@ impl fmt::Debug for Queued {
         )
     }
 }
+
+/// Equivalent to the command: `tar -xzf --strip-components 1`
+fn extract_tar_gz_data(data: impl Read, dest: impl AsRef<Path>) -> Result<()> {
+    let tar = GzDecoder::new(data);
+    let mut archive = Archive::new(tar);
+    archive.set_overwrite(true);
+
+    for entry in archive.entries()? {
+        let mut entry = entry?;
+        let path: PathBuf = entry.path()?.components().skip(1).collect();
+        entry.unpack(dest.as_ref().join(path))?;
+    }
+
+    Ok(())
+}
+
+/// Check for an existing marker file is the specified project directory and
+/// if one exists delete the indicated '.so' file.
+async fn remove_old_build(project_path: impl AsRef<Path>) -> Result<()> {
+    let marker_path = project_path.as_ref().join(MARKER_FILE_NAME);
+
+    if let Ok(mut existing_marker_file) = fs::File::open(&marker_path).await {
+        let mut old_so_name = String::new();
+        existing_marker_file
+            .read_to_string(&mut old_so_name)
+            .await?;
+
+        let old_so_path = project_path.as_ref().join(old_so_name);
+
+        if old_so_path.exists() {
+            fs::remove_file(old_so_path).await?;
+        }
+
+        fs::remove_file(marker_path).await?;
+    }
+
+    Ok(())
+}
+
+/// Give the '.so' file specified a random name so that re-deployments are
+/// properly re-loaded.
+async fn rename_build(project_path: impl AsRef<Path>, so_path: impl AsRef<Path>) -> Result<()> {
+    let random_so_name =
+        rand::distributions::Alphanumeric.sample_string(&mut rand::thread_rng(), 16);
+    let random_so_path = project_path.as_ref().join(&random_so_name);
+
+    fs::rename(so_path, random_so_path).await?;
+
+    fs::write(project_path.as_ref().join(MARKER_FILE_NAME), random_so_name).await?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use std::path::Path;
+    use tokio::fs;
+
+    use super::MARKER_FILE_NAME;
+
+    #[tokio::test]
+    async fn extract_tar_gz_data() {
+        let p = Path::new("/tmp/shuttle-extraction-test");
```

Comment:
We might want to use this crate to create temp directories instead :man_shrugging: 
https://docs.rs/tempdir